util/stmtsummary/statement_summary_test.go

```diff
@@ -882,3 +893,79 @@ func (s *testStmtSummarySuite) TestSummaryHistory(c *C) {
 	datum = s.ssMap.ToHistoryDatum()
 	c.Assert(len(datum), Equals, 5)
 }
+
+// Test summary when PrevSQL is not empty.
+func (s *testStmtSummarySuite) TestPrevSQL(c *C) {
+	s.ssMap.Clear()
+	now := time.Now().Unix()
+	// to disable expiration
+	s.ssMap.beginTimeForCurInterval = now + 60
+
+	stmtExecInfo1 := generateAnyExecInfo()
+	stmtExecInfo1.PrevSQL = "prevSQL"
+	stmtExecInfo1.PrevSQLDigest = "prevSQLDigest"
+	s.ssMap.AddStatement(stmtExecInfo1)
+	key := &stmtSummaryByDigestKey{
+		schemaName: stmtExecInfo1.SchemaName,
+		digest:     stmtExecInfo1.Digest,
+		prevDigest: stmtExecInfo1.PrevSQLDigest,
+	}
+	c.Assert(s.ssMap.summaryMap.Size(), Equals, 1)
+	_, ok := s.ssMap.summaryMap.Get(key)
+	c.Assert(ok, IsTrue)
+
+	// same prevSQL
+	s.ssMap.AddStatement(stmtExecInfo1)
+	c.Assert(s.ssMap.summaryMap.Size(), Equals, 1)
+
+	// different prevSQL
+	stmtExecInfo2 := stmtExecInfo1
+	stmtExecInfo2.PrevSQL = "prevSQL1"
+	stmtExecInfo2.PrevSQLDigest = "prevSQLDigest1"
+	key.prevDigest = stmtExecInfo2.PrevSQLDigest
+	s.ssMap.AddStatement(stmtExecInfo2)
+	c.Assert(s.ssMap.summaryMap.Size(), Equals, 2)
+	_, ok = s.ssMap.summaryMap.Get(key)
+	c.Assert(ok, IsTrue)
+}
+
+func (s *testStmtSummarySuite) TestEndTime(c *C) {
+	s.ssMap.Clear()
+	now := time.Now().Unix()
+	s.ssMap.beginTimeForCurInterval = now - 100
+
+	stmtExecInfo1 := generateAnyExecInfo()
+	s.ssMap.AddStatement(stmtExecInfo1)
+	key := &stmtSummaryByDigestKey{
+		schemaName: stmtExecInfo1.SchemaName,
+		digest:     stmtExecInfo1.Digest,
+	}
+	c.Assert(s.ssMap.summaryMap.Size(), Equals, 1)
+	value, ok := s.ssMap.summaryMap.Get(key)
+	c.Assert(ok, IsTrue)
+	ssbd := value.(*stmtSummaryByDigest)
+	ssElement := ssbd.history.Back().Value.(*stmtSummaryByDigestElement)
+	c.Assert(ssElement.beginTime, Equals, now-100)
+	c.Assert(ssElement.endTime, Equals, now+1700)
+
+	s.ssMap.SetRefreshInterval("3600", false)
+	defer s.ssMap.SetRefreshInterval("1800", false)
+	s.ssMap.AddStatement(stmtExecInfo1)
+	c.Assert(ssbd.history.Len(), Equals, 1)
+	ssElement = ssbd.history.Back().Value.(*stmtSummaryByDigestElement)
+	c.Assert(ssElement.beginTime, Equals, now-100)
+	c.Assert(ssElement.endTime, Equals, now+3500)
+
+	s.ssMap.SetRefreshInterval("60", false)
+	s.ssMap.AddStatement(stmtExecInfo1)
+	c.Assert(ssbd.history.Len(), Equals, 2)
+	now2 := time.Now().Unix()
+	ssElement = ssbd.history.Front().Value.(*stmtSummaryByDigestElement)
+	c.Assert(ssElement.beginTime, Equals, now-100)
+	c.Assert(ssElement.endTime, GreaterEqual, now)
+	c.Assert(ssElement.endTime, LessEqual, now2)
+	ssElement = ssbd.history.Back().Value.(*stmtSummaryByDigestElement)
+	c.Assert(ssElement.beginTime, GreaterEqual, now-60)
+	c.Assert(ssElement.beginTime, LessEqual, now2)
+	c.Assert(ssElement.endTime-ssElement.beginTime, Equals, int64(60))
```

Comment:
Do we need to add some cases to test the expiration?

Reply:
`TestSummaryHistory` tested expiration.